contracts/exchangeIssuance/ExchangeIssuanceLeveraged.sol

```diff
@@ -0,0 +1,926 @@
+     * @param _exchange               Exchange to use in swap from short to long token
+     */
+    function redeemExactSetForERC20(
+        ISetToken _setToken,
+        uint256 _amountSetToken,
+        address _outputToken,
+        uint256 _minAmountOutputToken,
+        Exchange _exchange
+    )
+        isSetToken(_setToken)
+        external
+        nonReentrant
+    {
+        bytes memory paymentParams = abi.encode(_minAmountOutputToken, _outputToken);
+        _initiateRedemption(_setToken, _amountSetToken, _exchange, PaymentToken.ERC20, paymentParams);
+    }
+
+    /**
+     * Trigger issuance of set token paying with any arbitrary ERC20 token
+     *
+     * @param _setToken            Set token to issue
+     * @param _amountSetToken      Amount to issue
+     * @param _maxAmountInputToken Maximum amount of input token to spend
+     * @param _exchange            Exchange to use in swap from short to long token
+     */
+    function issueExactSetFromERC20(
+        ISetToken _setToken,
+        uint256 _amountSetToken,
+        address _inputToken,
+        uint256 _maxAmountInputToken,
+        Exchange _exchange
+    )
+        isSetToken(_setToken)
+        external
+        nonReentrant
+    {
+        bytes memory paymentParams = abi.encode(_maxAmountInputToken, _inputToken);
+        _initiateIssuance(_setToken, _amountSetToken, _exchange, PaymentToken.ERC20, paymentParams);
+    }
+
+    /**
+     * Trigger issuance of set token paying with Eth
+     *
+     * @param _setToken            Set token to issue
+     * @param _amountSetToken      Amount to issue
+     * @param _exchange            Exchange to use in swap from short to long token
+     */
+    function issueExactSetFromETH(
+        ISetToken _setToken,
+        uint256 _amountSetToken,
+        Exchange _exchange
+    )
+        isSetToken(_setToken)
+        external
+        payable
+        nonReentrant
+    {
+        bytes memory paymentParams = abi.encode(msg.value);
+        _initiateIssuance(_setToken, _amountSetToken, _exchange, PaymentToken.ETH, paymentParams);
+    }
+
+    /**
+     * This is the callback function that will be called by the AaveLending Pool after flashloaned tokens have been sent
+     * to this contract.
+     * After exiting this function the Lending Pool will attempt to transfer back the loaned tokens + interest. If it fails to do so
+     * the whole transaction gets reverted
+     *
+     * @param assets     Addresses of all assets that were borrowed
+     * @param amounts    Amounts that were borrowed
+     * @param premiums   Interest to be paid on top of borrowed amount
+     * @param params     Encoded bytestring of other parameters from the original contract call to be used downstream
+     */
+    function executeOperation(
+        address[] calldata assets,
+        uint256[] calldata amounts,
+        uint256[] calldata premiums,
+        address , 
```

Comment:
Oh here's a really cool attack. We absolutely need to that this address (the initiator parameter) is our exchange issuance contract. Here's a scenario in which an attacker can ruin one of our users days:

Bob has previously used this contract to do a redemption and left an infinite approval of ETH2x open. Mallory creates a malicious contract that calls `flashloan` on Aave's contracts with the receiver address as this exchange issuance contract. Mallory makes sure to include malicious `params` that encodes for a redemption with an `originalSender` of Bob and a `minOutput` of 0. This will rip FLI out of Bob's wallet, but allow Mallory to sandwich attack any of the swaps done later in this transaction to steal of the money.

This scenario means anyone who has left an approval open to the contract can have their funds stolen by an attacker.

Reply:
Well spotted indeed. That would have been nasty. I will add a check that the initiator of the flashloan is the exchange issuance contract

Reply:
Added a check that prevents other people initiating flashloans on behalf of the EI contract. This should close above mentioned attack vector:
https://github.com/IndexCoop/index-coop-smart-contracts/pull/97/commits/e3933dd77055f4da79a02b55af78019d18f9aaf7